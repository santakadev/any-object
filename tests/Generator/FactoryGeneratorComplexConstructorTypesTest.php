<?php

declare(strict_types=1);

namespace Santakadev\AnyObject\Tests\Generator;

use ApprovalTests\Approvals;
use Santakadev\AnyObject\Tests\Generator\Generated\AnyNamedConstructorWithPrivateConstructObject;
use Santakadev\AnyObject\Tests\Generator\Generated\AnySharedTypesInConstructorObject;
use Santakadev\AnyObject\Tests\Generator\Generated\AnyVariadicNamedConstructorObject;
use Santakadev\AnyObject\Tests\TestData\ComplexConstructorTypes\NamedConstructorWithPrivateConstructObject;
use Santakadev\AnyObject\Tests\TestData\ComplexConstructorTypes\SharedTypesInConstructorObject;
use Santakadev\AnyObject\Tests\TestData\ComplexConstructorTypes\VariadicNamedConstructorObject;

class FactoryGeneratorComplexConstructorTypesTest extends FactoryGeneratorTestCase
{
    public function test_generator_from_named_constructor(): void
    {
        $this->generateFactoryFor(NamedConstructorWithPrivateConstructObject::class);

        $text = $this->readGeneratedAnyFileFor(NamedConstructorWithPrivateConstructObject::class);
        Approvals::verifyString($text);
        $test = AnyNamedConstructorWithPrivateConstructObject::with(value: "test");
        $this->assertEquals("test", $test->value);
    }

    public function test_generator_from_variadic_named_constructor(): void
    {
        $this->generateFactoryFor(VariadicNamedConstructorObject::class);

        $text = $this->readGeneratedAnyFileFor(VariadicNamedConstructorObject::class);
        Approvals::verifyString($text);
        $test = AnyVariadicNamedConstructorObject::with(value: ["a", "b"]);
        $this->assertEquals(["a", "b"], $test->value);
    }

    public function test_generator_shared_types_in_constructor(): void
    {
        $this->generateFactoryFor(SharedTypesInConstructorObject::class);

        $text = $this->readGeneratedAnyFileFor(SharedTypesInConstructorObject::class);
        Approvals::verifyString($text);
        AnySharedTypesInConstructorObject::build();
    }
}
