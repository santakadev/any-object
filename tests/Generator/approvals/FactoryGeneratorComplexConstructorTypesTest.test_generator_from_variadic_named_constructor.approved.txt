<?php

namespace Santakadev\AnyObject\Tests\Generator\Generated;

use Faker\Factory;
use Santakadev\AnyObject\Tests\TestData\ComplexConstructorTypes\VariadicNamedConstructorObject;

final class AnyVariadicNamedConstructorObject
{
    public static function with(array|None $value = new None()) : VariadicNamedConstructorObject
    {
        if ($value instanceof None) {
            $faker = Factory::create();
            $elementsCount = mt_rand(0, 50);
            $value = [];
            for ($i = 0; $i < $elementsCount; $i++) {
                $value[] = $faker->text(200);
            }
        }
        return VariadicNamedConstructorObject::fromString(...$value);
    }
    public static function build() : VariadicNamedConstructorObject
    {
        return self::with();
    }
}
